fdbserver/KeyValueStoreSQLite.actor.cpp

```diff
@@ -94,23 +95,40 @@ struct PageChecksumCodec {
 		SumType *pSumInPage = (SumType *)(pData + dataLen);
 
 		if (write) {
-			// Always write a CRC32 checksum for new pages
-			pSumInPage->part1 = 0; // Indicates CRC32 is being used
-			pSumInPage->part2 = crc32c_append(0xfdbeefdb, static_cast<uint8_t*>(data), dataLen);
+			// Always write a xxhash64 checksum for new pages
+			// First 8 bits are set to 0 so that with high probability,
+			// checksums written with hashlittle2 don't require calculating
+			// an xxHash64 checksum on read
+			auto xxHash64 = XXHash64::hash(data, dataLen, 0xfdbeefdbfdbeefdb);
+			pSumInPage->part1 = static_cast<uint32_t>((xxHash64 >> 32) & 0x00ffffff);
+			pSumInPage->part2 = static_cast<uint32_t>(xxHash64 & 0xffffffff);
 			return true;
 		}
 
-		SumType sum;
+		SumType crc32Sum;
 		if (pSumInPage->part1 == 0) {
-			// part1 being 0 indicates with high probability that a CRC32 checksum
+			// part1 being 0 indicates with very high probability that a CRC32 checksum
 			// was used, so check that first. If this checksum fails, there is still
-			// some chance the page was written with hashlittle2, so fall back to checking
+			// some chance the page was written with another checksum algorithm
+			crc32Sum.part1 = 0;
+			crc32Sum.part2 = crc32c_append(0xfdbeefdb, static_cast<uint8_t*>(data), dataLen);
+			if (crc32Sum == *pSumInPage) return true;
+		}
+
+		// Try xxhash64
+		SumType xxHash64Sum;
+		if ((pSumInPage->part1 >> 24) == 0) {
+			// The first 8 bits of part1 being 0 indicates with high probability that an
+			// xxHash64 checksum was used, so check that next. If this checksum fails, there is
+			// still some change the page was written with hashlittle2, so fall back to checking
 			// hashlittle2
-			sum.part1 = 0;
-			sum.part2 = crc32c_append(0xfdbeefdb, static_cast<uint8_t*>(data), dataLen);
-			if (sum == *pSumInPage) return true;
+			auto xxHash64 = XXHash64::hash(data, dataLen, 0xfdbeefdbfdbeefdb);
```

Comment:
According to the limitations section on https://create.stephan-brumme.com/xxhash/

> When fed with unaligned data, it could be MUCH slower than the reference implementation.

are we sure this data is always aligned? (I assume they mean aligned to 8 bytes but I'm not 100% sure)

For #3732 , we'll need it to be able to handle unaligned data so I'm wondering if we should bring in the reference implementation.

Reply:
Also I'm pretty sure this is xxhash64, not xxhash3

Reply:
> When fed with unaligned data, it could be MUCH slower than the reference implementation.

I ran some microbenchmarks and didn't notice a significant performance decrease when memory was unaligned.

Reply:
> Also I'm pretty sure this is xxhash64, not xxhash3

@sfc-gh-mpilman Can you confirm if xxhash64 is ok instead of xxhash3?

Reply:
I think alignment in sqlite should be fine. What is the rational for using xxhash64?

Reply:
I incorrectly thought xxhash3 was still experimental, but I see it's stable now. I tested the performance with `flowbench` and the results are consistent with http://fastcompression.blogspot.com/2019/03/presenting-xxh3.html, so I switched to xxhash3.